File: mac/Core/Protocol/ControlFrame.swift
import Foundation

public enum ControlType: UInt8, Equatable {
    case hello = 0x01
    case layoutSync = 0x02
    case roleSwitch = 0x03
    case ownershipTransfer = 0x04
    case screenAnnounce = 0x05
    case wakeOnLanRequest = 0x06
    case hotkey = 0x07
    case clipboardOffer = 0x10
    case clipboardRequest = 0x11
    case clipboardData = 0x12
    case fileOffer = 0x20
    case fileChunk = 0x21
    case fileAck = 0x22
    case fileEnd = 0x23
    case keepAlive = 0xF0
    case bye = 0xF1
}

public struct ControlFrame: Equatable {
    public let type: ControlType
    public let body: Data

    public init(type: ControlType, body: Data = Data()) {
        self.type = type
        self.body = body
    }
}

public enum ControlFrameCodec {
    /// Maximum frame body size accepted by the decoder. Prevents huge-allocation DoS.
    /// Files and clipboards above this go on dedicated sub-streams chunked accordingly.
    public static let maxBodySize: Int = 16 * 1024 * 1024

    public static func encode(_ frame: ControlFrame) -> Data {
        let bodyLen = frame.body.count
        precondition(bodyLen <= maxBodySize, "ControlFrame body exceeds max size")
        let length = UInt32(1 + bodyLen)
        var data = Data(capacity: 4 + Int(length))
        data.appendLE(length)
        data.append(frame.type.rawValue)
        data.append(frame.body)
        return data
    }

    /// Decode one frame from a buffer. Returns the frame and the number of bytes consumed.
    /// If `data` does not yet contain a complete frame, throws `truncated` (caller should
    /// read more bytes and retry).
    public static func decode(_ data: Data) throws -> (frame: ControlFrame, consumed: Int) {
        guard data.count >= 4 else { throw BrpCodecError.truncated }
        let length: UInt32 = data.readLE(at: 0)
        guard length >= 1 else { throw BrpCodecError.truncated }
        let bodyLen = Int(length) - 1
        guard bodyLen <= maxBodySize else { throw BrpCodecError.lengthOverflow }
        let totalSize = 4 + Int(length)
        guard data.count >= totalSize else { throw BrpCodecError.truncated }
        let typeRaw = data.readU8(at: 4)
        guard let type = ControlType(rawValue: typeRaw) else {
            throw BrpCodecError.unknownType(typeRaw)
        }
        let bodyStart = data.startIndex + 5
        let bodyEnd = data.startIndex + totalSize
        let body = Data(data[bodyStart..<bodyEnd])
        return (ControlFrame(type: type, body: body), totalSize)
    }
}

File: windows/Core/Protocol/ControlFrame.cs
using System.Buffers.Binary;

namespace BarelyReal.Core.Protocol;

public enum ControlType : byte
{
    Hello = 0x01,
    LayoutSync = 0x02,
    RoleSwitch = 0x03,
    OwnershipTransfer = 0x04,
    ScreenAnnounce = 0x05,
    WakeOnLanRequest = 0x06,
    Hotkey = 0x07,
    ClipboardOffer = 0x10,
    ClipboardRequest = 0x11,
    ClipboardData = 0x12,
    FileOffer = 0x20,
    FileChunk = 0x21,
    FileAck = 0x22,
    FileEnd = 0x23,
    KeepAlive = 0xF0,
    Bye = 0xF1
}

public sealed class ControlFrame : IEquatable<ControlFrame>
{
    public ControlType Type { get; }
    public byte[] Body { get; }

    public ControlFrame(ControlType type, byte[]? body = null)
    {
        Type = type;
        Body = body ?? Array.Empty<byte>();
    }

    public bool Equals(ControlFrame? other)
    {
        if (other is null) return false;
        return Type == other.Type && Body.AsSpan().SequenceEqual(other.Body.AsSpan());
    }

    public override bool Equals(object? obj) => obj is ControlFrame other && Equals(other);
    public override int GetHashCode() => HashCode.Combine(Type, Body.Length);
}

public static class ControlFrameCodec
{
    /// Maximum frame body size accepted by the decoder. Prevents huge-allocation DoS.
    public const int MaxBodySize = 16 * 1024 * 1024;

    public static byte[] Encode(ControlFrame frame)
    {
        if (frame.Body.Length > MaxBodySize)
            throw new ArgumentException($"Body exceeds {MaxBodySize}", nameof(frame));

        var length = (uint)(1 + frame.Body.Length);
        var buf = new byte[4 + length];
        var span = buf.AsSpan();
        BinaryPrimitives.WriteUInt32LittleEndian(span[..4], length);
        span[4] = (byte)frame.Type;
        if (frame.Body.Length > 0)
        {
            frame.Body.AsSpan().CopyTo(span[5..]);
        }
        return buf;
    }

    /// Decode one frame from a buffer. Returns the frame and the number of bytes consumed.
    /// Throws <see cref="BrpCodecException"/> with <see cref="BrpCodecError.Truncated"/> if a
    /// complete frame is not yet available — caller should read more bytes and retry.
    public static (ControlFrame Frame, int Consumed) Decode(ReadOnlySpan<byte> data)
    {
        if (data.Length < 4)
            throw new BrpCodecException(BrpCodecError.Truncated);

        var length = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        if (length < 1)
            throw new BrpCodecException(BrpCodecError.Truncated);

        var bodyLen = (int)length - 1;
        if (bodyLen > MaxBodySize)
            throw new BrpCodecException(BrpCodecError.LengthOverflow);

        var totalSize = 4 + (int)length;
        if (data.Length < totalSize)
            throw new BrpCodecException(BrpCodecError.Truncated);

        var typeRaw = data[4];
        if (!Enum.IsDefined(typeof(ControlType), typeRaw))
            throw new BrpCodecException(BrpCodecError.UnknownType, typeRaw);

        var body = data.Slice(5, bodyLen).ToArray();
        return (new ControlFrame((ControlType)typeRaw, body), totalSize);
    }
}
